File: Trackfield/app/src/main/java/com/felwal/trackfield/ui/widget/dialog/RadioDialog.java
package com.felwal.trackfield.ui.widget.dialog;

import android.app.AlertDialog;
import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.RadioButton;
import android.widget.RadioGroup;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

import com.felwal.trackfield.R;

import java.util.ArrayList;

public class RadioDialog extends BaseDialog {

    // bundle keys
    public static final String BUNDLE_RADIO_TEXTS = "radioTexts";
    public static final String BUNDLE_SELECTED_INDEX = "selectedIndex";

    // dialog tags
    public static final String TAG_DEFAULT = "radioDialog";

    private DialogListener listener;

    // arguments
    private ArrayList<String> radioTexts;
    private int selectedIndex;

    //

    public static RadioDialog newInstance(@StringRes int titleRes, @StringRes int messageRes,
        ArrayList<String> radioButtonTexts, int selectedIndex, String tag) {

        RadioDialog instance = new RadioDialog();
        Bundle bundle = putBundleBase(titleRes, messageRes, NO_RES, tag);

        bundle.putStringArrayList(BUNDLE_RADIO_TEXTS, radioButtonTexts);
        bundle.putInt(BUNDLE_SELECTED_INDEX, selectedIndex);

        instance.setArguments(bundle);
        return instance;
    }

    // on

    @Override
    public void onAttach(@NonNull Context c) {
        super.onAttach(c);

        try {
            listener = (DialogListener) c;
        }
        catch (ClassCastException e) {
            throw new ClassCastException("Activity must implement DialogListener");
        }
    }

    // extends

    @Override
    protected void unpackBundle() {
        Bundle bundle = unpackBundleBase(TAG_DEFAULT);

        if (bundle != null) {
            radioTexts = bundle.getStringArrayList(BUNDLE_RADIO_TEXTS);
            selectedIndex = bundle.getInt(BUNDLE_SELECTED_INDEX, 0);
        }

        // keep in range
        if (selectedIndex >= radioTexts.size()) selectedIndex = radioTexts.size() - 1;
        else if (selectedIndex < 0) selectedIndex = 0;
    }

    @Override
    protected AlertDialog buildDialog() {
        final View view = inflater.inflate(R.layout.dialog_radio, null);
        final RadioGroup group = view.findViewById(R.id.rg_dialog_radio);

        builder.setView(view).setTitle(title)
            .setNegativeButton(negBtnTxtRes, (dialog, id) -> getDialog().cancel());

        // inflate radio buttons
        for (int i = 0; i < radioTexts.size(); i++) {
            RadioButton btn = (RadioButton) inflater.inflate(R.layout.item_dialog_radio, group, false);
            btn.setText(radioTexts.get(i));
            group.addView(btn);

            if (i == selectedIndex) btn.setChecked(true);

            // click
            int iFinal = i;
            btn.setOnClickListener(v -> {
                getDialog().cancel();
                listener.onRadioDialogClick(iFinal, tag);
            });
        }

        return builder.show();
    }

    // interface

    public interface DialogListener {

        void onRadioDialogClick(int index, String tag);

    }

}

File: Trackfield/android/src/main/java/com/felwal/android/widget/dialog/RadioDialog.kt
package com.felwal.android.widget.dialog

import android.content.Context
import android.os.Bundle
import androidx.annotation.StringRes
import androidx.appcompat.app.AlertDialog
import com.felwal.android.R

private const val ARG_ITEMS = "items"
private const val ARG_CHECKED_ITEM = "checkedItem"

class RadioDialog : BaseDialog() {

    private lateinit var listener: DialogListener

    // args
    private lateinit var items: Array<out String>
    private var checkedItem = 0

    // DialogFragment

    override fun onAttach(c: Context) {
        super.onAttach(c)

        listener = try {
            c as DialogListener
        }
        catch (e: ClassCastException) {
            throw ClassCastException("Activity must implement DialogListener")
        }
    }

    // BaseDialog

    override fun unpackBundle(bundle: Bundle?) {
        bundle?.apply {
            items = getStringArray(ARG_ITEMS).orEmpty()
            checkedItem = getInt(ARG_CHECKED_ITEM, 0).coerceIn(0, items.size)
        }
    }

    override fun buildDialog(): AlertDialog = builder.run {
        setTitle(title)
        if (message != "") setMessage(message)

        setSingleChoiceItems(items, checkedItem) { dialog, which ->
            dialog.cancel()
            listener.onRadioDialogItemClick(which, dialogTag)
        }
        setCancelButton(negBtnTxtRes)

        show()
    }

    //

    interface DialogListener {
        fun onRadioDialogItemClick(checkedItem: Int, tag: String)
    }

    //

    companion object {
        @JvmStatic
        fun newInstance(
            title: String,
            message: String = "",
            items: List<String>,
            checkedItem: Int,
            @StringRes negBtnTxtRes: Int = R.string.dialog_btn_cancel,
            tag: String
        ): RadioDialog = RadioDialog().apply {
            arguments = putBaseBundle(title, message, NO_RES, negBtnTxtRes = negBtnTxtRes, tag = tag).apply {
                putStringArray(ARG_ITEMS, items.toTypedArray())
                putInt(ARG_CHECKED_ITEM, checkedItem)
            }
        }
    }
}

fun radioDialog(
    title: String,
    message: String = "",
    items: List<String>,
    checkedItem: Int,
    @StringRes negBtnTxtRes: Int = R.string.dialog_btn_cancel,
    tag: String
): RadioDialog = RadioDialog.newInstance(title, message, items, checkedItem, negBtnTxtRes, tag)